osu.Game/Rulesets/Ruleset.cs

```diff
@@ -212,7 +212,19 @@ public virtual LegacyMods ConvertToLegacyMods(Mod[] mods)
         /// <param name="skin">The source skin.</param>
         /// <param name="beatmap">The current beatmap.</param>
         /// <returns>A skin with a transformer applied, or null if no transformation is provided by this ruleset.</returns>
-        public virtual ISkin? CreateSkinTransformer(ISkin skin, IBeatmap beatmap) => null;
+        public virtual ISkin? CreateSkinTransformer(ISkin skin, IBeatmap beatmap)
+        {
+            switch (skin)
+            {
+                case LegacySkin:
+                    return new LegacySkinTransformer(skin);
+
+                case ArgonSkin:
+                    return new ArgonSkinTransformer(skin);
+            }
```

Comment:
Sounds about right, I'm following how the API is right now. I think it would look better if we split each type of skin into a separate method, e.g. `Create{Argon,Triangles,Legacy}SkinTransformer()`, to make the point of "extending default skin to custom ruleset" clearer.

Reply:
<details>
<summary>diff:</summary>

```diff
diff --git a/osu.Game.Rulesets.Catch/CatchRuleset.cs b/osu.Game.Rulesets.Catch/CatchRuleset.cs
index 72d1a161dd..42538f10f7 100644
--- a/osu.Game.Rulesets.Catch/CatchRuleset.cs
+++ b/osu.Game.Rulesets.Catch/CatchRuleset.cs
@@ -199,19 +199,9 @@ public override LocalisableString GetDisplayNameForHitResult(HitResult result)
 
         public override DifficultyCalculator CreateDifficultyCalculator(IWorkingBeatmap beatmap) => new CatchDifficultyCalculator(RulesetInfo, beatmap);
 
-        public override ISkin? CreateSkinTransformer(ISkin skin, IBeatmap beatmap)
-        {
-            switch (skin)
-            {
-                case LegacySkin:
-                    return new CatchLegacySkinTransformer(skin);
+        public override ISkin CreateArgonSkinTransformer(ArgonSkin skin, IBeatmap beatmap) => new CatchArgonSkinTransformer(skin);
 
-                case ArgonSkin:
-                    return new CatchArgonSkinTransformer(skin);
-            }
-
-            return null;
-        }
+        public override ISkin CreateLegacySkinTransformer(LegacySkin skin, IBeatmap beatmap) => new CatchLegacySkinTransformer(skin);
 
         public override PerformanceCalculator CreatePerformanceCalculator() => new CatchPerformanceCalculator();
 
diff --git a/osu.Game.Rulesets.Catch/Skinning/Argon/CatchArgonSkinTransformer.cs b/osu.Game.Rulesets.Catch/Skinning/Argon/CatchArgonSkinTransformer.cs
index a67945df98..1c3f960f9a 100644
--- a/osu.Game.Rulesets.Catch/Skinning/Argon/CatchArgonSkinTransformer.cs
+++ b/osu.Game.Rulesets.Catch/Skinning/Argon/CatchArgonSkinTransformer.cs
@@ -8,7 +8,7 @@ namespace osu.Game.Rulesets.Catch.Skinning.Argon
 {
     public class CatchArgonSkinTransformer : ArgonSkinTransformer
     {
-        public CatchArgonSkinTransformer(ISkin skin)
+        public CatchArgonSkinTransformer(ArgonSkin skin)
             : base(skin)
         {
         }
diff --git a/osu.Game.Rulesets.Catch/Skinning/Legacy/CatchLegacySkinTransformer.cs b/osu.Game.Rulesets.Catch/Skinning/Legacy/CatchLegacySkinTransformer.cs
index 462fd5ab64..b689bf41b4 100644
--- a/osu.Game.Rulesets.Catch/Skinning/Legacy/CatchLegacySkinTransformer.cs
+++ b/osu.Game.Rulesets.Catch/Skinning/Legacy/CatchLegacySkinTransformer.cs
@@ -20,7 +20,7 @@ public class CatchLegacySkinTransformer : LegacySkinTransformer
         /// </summary>
         private bool providesComboCounter => this.HasFont(LegacyFont.Combo);
 
-        public CatchLegacySkinTransformer(ISkin skin)
+        public CatchLegacySkinTransformer(LegacySkin skin)
             : base(skin)
         {
         }
diff --git a/osu.Game.Rulesets.Mania/ManiaRuleset.cs b/osu.Game.Rulesets.Mania/ManiaRuleset.cs
index b5614e2b56..e3a0925cc8 100644
--- a/osu.Game.Rulesets.Mania/ManiaRuleset.cs
+++ b/osu.Game.Rulesets.Mania/ManiaRuleset.cs
@@ -67,24 +67,16 @@ public class ManiaRuleset : Ruleset, ILegacyRuleset
 
         public override IBeatmapVerifier CreateBeatmapVerifier() => new ManiaBeatmapVerifier();
 
-        public override ISkin? CreateSkinTransformer(ISkin skin, IBeatmap beatmap)
-        {
-            switch (skin)
-            {
-                case TrianglesSkin:
-                    return new ManiaTrianglesSkinTransformer(skin, beatmap);
-
-                case ArgonSkin:
-                    return new ManiaArgonSkinTransformer(skin, beatmap);
+        public override ISkin CreateArgonSkinTransformer(ArgonSkin skin, IBeatmap beatmap) => new ManiaArgonSkinTransformer(skin, beatmap);
 
-                case DefaultLegacySkin:
-                    return new ManiaClassicSkinTransformer(skin, beatmap);
+        public override ISkin CreateTrianglesSkinTransformer(TrianglesSkin skin, IBeatmap beatmap) => new ManiaTrianglesSkinTransformer(skin, beatmap);
 
-                case LegacySkin:
-                    return new ManiaLegacySkinTransformer(skin, beatmap);
-            }
+        public override ISkin CreateLegacySkinTransformer(LegacySkin skin, IBeatmap beatmap)
+        {
+            if (skin is DefaultLegacySkin)
+                return new ManiaClassicSkinTransformer(skin, beatmap);
 
-            return null;
+            return new ManiaLegacySkinTransformer(skin, beatmap);
         }
 
         public override IEnumerable<Mod> ConvertFromLegacyMods(LegacyMods mods)
diff --git a/osu.Game.Rulesets.Mania/Skinning/Argon/ManiaArgonSkinTransformer.cs b/osu.Game.Rulesets.Mania/Skinning/Argon/ManiaArgonSkinTransformer.cs
index 7f6540e7b5..5a22664354 100644
--- a/osu.Game.Rulesets.Mania/Skinning/Argon/ManiaArgonSkinTransformer.cs
+++ b/osu.Game.Rulesets.Mania/Skinning/Argon/ManiaArgonSkinTransformer.cs
@@ -12,11 +12,11 @@
 
 namespace osu.Game.Rulesets.Mania.Skinning.Argon
 {
-    public class ManiaArgonSkinTransformer : SkinTransformer
+    public class ManiaArgonSkinTransformer : ArgonSkinTransformer
     {
         private readonly ManiaBeatmap beatmap;
 
-        public ManiaArgonSkinTransformer(ISkin skin, IBeatmap beatmap)
+        public ManiaArgonSkinTransformer(ArgonSkin skin, IBeatmap beatmap)
             : base(skin)
         {
             this.beatmap = (ManiaBeatmap)beatmap;
diff --git a/osu.Game.Rulesets.Mania/Skinning/Default/ManiaTrianglesSkinTransformer.cs b/osu.Game.Rulesets.Mania/Skinning/Default/ManiaTrianglesSkinTransformer.cs
index 3e0fe8ed4b..33da38b3fb 100644
--- a/osu.Game.Rulesets.Mania/Skinning/Default/ManiaTrianglesSkinTransformer.cs
+++ b/osu.Game.Rulesets.Mania/Skinning/Default/ManiaTrianglesSkinTransformer.cs
@@ -14,7 +14,7 @@ public class ManiaTrianglesSkinTransformer : SkinTransformer
     {
         private readonly ManiaBeatmap beatmap;
 
-        public ManiaTrianglesSkinTransformer(ISkin skin, IBeatmap beatmap)
+        public ManiaTrianglesSkinTransformer(TrianglesSkin skin, IBeatmap beatmap)
             : base(skin)
         {
             this.beatmap = (ManiaBeatmap)beatmap;
diff --git a/osu.Game.Rulesets.Mania/Skinning/Legacy/ManiaClassicSkinTransformer.cs b/osu.Game.Rulesets.Mania/Skinning/Legacy/ManiaClassicSkinTransformer.cs
index be3372fe58..00739595b2 100644
--- a/osu.Game.Rulesets.Mania/Skinning/Legacy/ManiaClassicSkinTransformer.cs
+++ b/osu.Game.Rulesets.Mania/Skinning/Legacy/ManiaClassicSkinTransformer.cs
@@ -10,7 +10,7 @@ namespace osu.Game.Rulesets.Mania.Skinning.Legacy
 {
     public class ManiaClassicSkinTransformer : ManiaLegacySkinTransformer
     {
-        public ManiaClassicSkinTransformer(ISkin skin, IBeatmap beatmap)
+        public ManiaClassicSkinTransformer(LegacySkin skin, IBeatmap beatmap)
             : base(skin, beatmap)
         {
         }
diff --git a/osu.Game.Rulesets.Mania/Skinning/Legacy/ManiaLegacySkinTransformer.cs b/osu.Game.Rulesets.Mania/Skinning/Legacy/ManiaLegacySkinTransformer.cs
index 73c521b2ed..769135a090 100644
--- a/osu.Game.Rulesets.Mania/Skinning/Legacy/ManiaLegacySkinTransformer.cs
+++ b/osu.Game.Rulesets.Mania/Skinning/Legacy/ManiaLegacySkinTransformer.cs
@@ -61,7 +61,7 @@ private static readonly IReadOnlyDictionary<HitResult, string> default_hit_resul
 
         private readonly ManiaBeatmap beatmap;
 
-        public ManiaLegacySkinTransformer(ISkin skin, IBeatmap beatmap)
+        public ManiaLegacySkinTransformer(LegacySkin skin, IBeatmap beatmap)
             : base(skin)
         {
             this.beatmap = (ManiaBeatmap)beatmap;
diff --git a/osu.Game.Rulesets.Osu/OsuRuleset.cs b/osu.Game.Rulesets.Osu/OsuRuleset.cs
index 6752712be1..9061c601a1 100644
--- a/osu.Game.Rulesets.Osu/OsuRuleset.cs
+++ b/osu.Game.Rulesets.Osu/OsuRuleset.cs
@@ -246,22 +246,11 @@ public override IEnumerable<Mod> GetModsFor(ModType type)
 
         public override RulesetSettingsSubsection CreateSettings() => new OsuSettingsSubsection(this);
 
-        public override ISkin? CreateSkinTransformer(ISkin skin, IBeatmap beatmap)
-        {
-            switch (skin)
-            {
-                case LegacySkin:
-                    return new OsuLegacySkinTransformer(skin);
+        public override ISkin CreateLegacySkinTransformer(LegacySkin skin, IBeatmap beatmap) => new OsuLegacySkinTransformer(skin);
 
-                case ArgonSkin:
-                    return new OsuArgonSkinTransformer(skin);
+        public override ISkin CreateArgonSkinTransformer(ArgonSkin skin, IBeatmap beatmap) => new OsuArgonSkinTransformer(skin);
 
-                case TrianglesSkin:
-                    return new OsuTrianglesSkinTransformer(skin);
-            }
-
-            return null;
-        }
+        public override ISkin CreateTrianglesSkinTransformer(TrianglesSkin skin, IBeatmap beatmap) => new OsuTrianglesSkinTransformer(skin);
 
         public int LegacyID => 0;
 
diff --git a/osu.Game.Rulesets.Taiko/Skinning/Argon/TaikoArgonSkinTransformer.cs b/osu.Game.Rulesets.Taiko/Skinning/Argon/TaikoArgonSkinTransformer.cs
index 7d38d6c9e5..e2dc7cf3ba 100644
--- a/osu.Game.Rulesets.Taiko/Skinning/Argon/TaikoArgonSkinTransformer.cs
+++ b/osu.Game.Rulesets.Taiko/Skinning/Argon/TaikoArgonSkinTransformer.cs
@@ -9,7 +9,7 @@ namespace osu.Game.Rulesets.Taiko.Skinning.Argon
 {
     public class TaikoArgonSkinTransformer : ArgonSkinTransformer
     {
-        public TaikoArgonSkinTransformer(ISkin skin)
+        public TaikoArgonSkinTransformer(ArgonSkin skin)
             : base(skin)
         {
         }
diff --git a/osu.Game.Rulesets.Taiko/Skinning/Legacy/TaikoLegacySkinTransformer.cs b/osu.Game.Rulesets.Taiko/Skinning/Legacy/TaikoLegacySkinTransformer.cs
index 894b91e9ce..ec9c571edf 100644
--- a/osu.Game.Rulesets.Taiko/Skinning/Legacy/TaikoLegacySkinTransformer.cs
+++ b/osu.Game.Rulesets.Taiko/Skinning/Legacy/TaikoLegacySkinTransformer.cs
@@ -21,7 +21,7 @@ public class TaikoLegacySkinTransformer : LegacySkinTransformer
         private bool hasHitCircle => GetTexture("taikohitcircle") != null;
         private bool hasBarLeft => GetTexture("taiko-bar-left") != null;
 
-        public TaikoLegacySkinTransformer(ISkin skin)
+        public TaikoLegacySkinTransformer(LegacySkin skin)
             : base(skin)
         {
             hasExplosion = new Lazy<bool>(() => GetTexture(getHitName(TaikoSkinComponents.TaikoExplosionGreat)) != null);
diff --git a/osu.Game.Rulesets.Taiko/TaikoRuleset.cs b/osu.Game.Rulesets.Taiko/TaikoRuleset.cs
index 13501c6192..f137d64f8f 100644
--- a/osu.Game.Rulesets.Taiko/TaikoRuleset.cs
+++ b/osu.Game.Rulesets.Taiko/TaikoRuleset.cs
@@ -49,19 +49,9 @@ public class TaikoRuleset : Ruleset, ILegacyRuleset
 
         public override IBeatmapConverter CreateBeatmapConverter(IBeatmap beatmap) => new TaikoBeatmapConverter(beatmap, this);
 
-        public override ISkin? CreateSkinTransformer(ISkin skin, IBeatmap beatmap)
-        {
-            switch (skin)
-            {
-                case ArgonSkin:
-                    return new TaikoArgonSkinTransformer(skin);
+        public override ISkin CreateLegacySkinTransformer(LegacySkin skin, IBeatmap beatmap) => new TaikoLegacySkinTransformer(skin);
 
-                case LegacySkin:
-                    return new TaikoLegacySkinTransformer(skin);
-            }
-
-            return null;
-        }
+        public override ISkin CreateArgonSkinTransformer(ArgonSkin skin, IBeatmap beatmap) => new TaikoArgonSkinTransformer(skin);
 
         public const string SHORT_NAME = "taiko";
 
diff --git a/osu.Game/Rulesets/Ruleset.cs b/osu.Game/Rulesets/Ruleset.cs
index c7d4779064..ca5894e5c2 100644
--- a/osu.Game/Rulesets/Ruleset.cs
+++ b/osu.Game/Rulesets/Ruleset.cs
@@ -207,24 +207,28 @@ public virtual LegacyMods ConvertToLegacyMods(Mod[] mods)
         public ModTouchDevice? GetTouchDeviceMod() => CreateMod<ModTouchDevice>();
 
         /// <summary>
-        /// Create a transformer which adds lookups specific to a ruleset to skin sources.
+        /// Creates an <see cref="ArgonSkin"/>-based transformer which adds lookups specific to a ruleset to skin sources.
         /// </summary>
         /// <param name="skin">The source skin.</param>
         /// <param name="beatmap">The current beatmap.</param>
         /// <returns>A skin with a transformer applied, or null if no transformation is provided by this ruleset.</returns>
-        public virtual ISkin? CreateSkinTransformer(ISkin skin, IBeatmap beatmap)
-        {
-            switch (skin)
-            {
-                case LegacySkin:
-                    return new LegacySkinTransformer(skin);
+        public virtual ISkin? CreateArgonSkinTransformer(ArgonSkin skin, IBeatmap beatmap) => new ArgonSkinTransformer(skin);
 
-                case ArgonSkin:
-                    return new ArgonSkinTransformer(skin);
-            }
+        /// <summary>
+        /// Creates a <see cref="TrianglesSkin"/>-based transformer which adds lookups specific to a ruleset to skin sources.
+        /// </summary>
+        /// <param name="skin">The source skin.</param>
+        /// <param name="beatmap">The current beatmap.</param>
+        /// <returns>A skin with a transformer applied, or null if no transformation is provided by this ruleset.</returns>
+        public virtual ISkin? CreateTrianglesSkinTransformer(TrianglesSkin skin, IBeatmap beatmap) => null;
 
-            return null;
-        }
+        /// <summary>
+        /// Creates a <see cref="LegacySkin"/>-based transformer which adds lookups specific to a ruleset to skin sources.
+        /// </summary>
+        /// <param name="skin">The source skin.</param>
+        /// <param name="beatmap">The current beatmap.</param>
+        /// <returns>A skin with a transformer applied, or null if no transformation is provided by this ruleset.</returns>
+        public virtual ISkin? CreateLegacySkinTransformer(LegacySkin skin, IBeatmap beatmap) => new LegacySkinTransformer(skin);
 
         protected Ruleset()
         {
diff --git a/osu.Game/Skinning/ArgonSkinTransformer.cs b/osu.Game/Skinning/ArgonSkinTransformer.cs
index 387a7a9c0b..304f64aca7 100644
--- a/osu.Game/Skinning/ArgonSkinTransformer.cs
+++ b/osu.Game/Skinning/ArgonSkinTransformer.cs
@@ -10,7 +10,7 @@ namespace osu.Game.Skinning
 {
     public class ArgonSkinTransformer : SkinTransformer
     {
-        public ArgonSkinTransformer(ISkin skin)
+        public ArgonSkinTransformer(ArgonSkin skin)
             : base(skin)
         {
         }
diff --git a/osu.Game/Skinning/LegacySkinTransformer.cs b/osu.Game/Skinning/LegacySkinTransformer.cs
index 3ea316c0c7..30bb4f2074 100644
--- a/osu.Game/Skinning/LegacySkinTransformer.cs
+++ b/osu.Game/Skinning/LegacySkinTransformer.cs
@@ -18,7 +18,7 @@ public class LegacySkinTransformer : SkinTransformer
         /// </summary>
         public virtual bool IsProvidingLegacyResources => this.HasFont(LegacyFont.Combo);
 
-        public LegacySkinTransformer(ISkin skin)
+        public LegacySkinTransformer(LegacySkin skin)
             : base(skin)
         {
         }
diff --git a/osu.Game/Skinning/RulesetSkinProvidingContainer.cs b/osu.Game/Skinning/RulesetSkinProvidingContainer.cs
index 07e238243b..8d4268a49e 100644
--- a/osu.Game/Skinning/RulesetSkinProvidingContainer.cs
+++ b/osu.Game/Skinning/RulesetSkinProvidingContainer.cs
@@ -100,9 +100,17 @@ protected ISkin GetRulesetTransformedSkin(ISkin skin)
             if (skin == null)
                 return null;
 
-            var rulesetTransformed = Ruleset.CreateSkinTransformer(skin, Beatmap);
-            if (rulesetTransformed != null)
-                return rulesetTransformed;
+            switch (skin)
+            {
+                case ArgonSkin argon:
+                    return Ruleset.CreateArgonSkinTransformer(argon, Beatmap) ?? argon;
+
+                case TrianglesSkin triangles:
+                    return Ruleset.CreateTrianglesSkinTransformer(triangles, Beatmap) ?? triangles;
+
+                case LegacySkin legacy:
+                    return Ruleset.CreateLegacySkinTransformer(legacy, Beatmap) ?? legacy;
+            }
 
             return skin;
         }
```

</details>

Reply:
Let's leave that patch out of this diff. I have mixed feelings on it in general.